kernel/kernel-api/src/main/java/io/delta/kernel/internal/commit/CommitContextImpl.java

```diff
@@ -0,0 +1,155 @@
+  }
+
+  // TODO: forRebasedCommit
+
+  //////////////////////////////////
+  // Member variables and methods //
+  //////////////////////////////////
+
+  private final TransactionV2State txnState;
+  private final CloseableIterator<Row> finalizedDataActions;
+  private final long commitAttemptTimestampMs;
+  private final CommitInfo commitInfo;
+  private final Metadata metadata;
+  private boolean iteratorConsumed;
+
+  private CommitContextImpl(
+      Engine engine, TransactionV2State txnState, CloseableIterator<Row> finalizedDataActions) {
+    this.txnState = txnState;
+    this.finalizedDataActions = finalizedDataActions;
+    this.commitAttemptTimestampMs = txnState.clock.getTimeMillis();
+    this.metadata = txnState.initialMetadata; // TODO: update with ICT enablement info on conflict
+    this.commitInfo = getCommitInfo();
+    this.iteratorConsumed = false;
+  }
+
+  /////////////////
+  // Public APIs //
+  /////////////////
+
+  @Override
+  public CloseableIterator<Row> getFinalizedActions() {
+    checkState(!iteratorConsumed, "Finalized actions iterator has already been consumed.");
+    iteratorConsumed = true;
+    return getMetadataActions().combine(finalizedDataActions);
+  }
+
+  @Override
+  public CommitMetadata getCommitMetadata() {
+    return new CommitMetadata(
+        getCommitAsVersion(),
+        txnState.logPath,
+        commitInfo,
+        txnState.readTableOpt.map(TransactionDataSource::getProtocol),
+        txnState.readTableOpt.map(TransactionDataSource::getMetadata),
+        txnState.isProtocolUpdate ? Optional.of(txnState.protocol) : Optional.empty(),
+        txnState.isMetadataUpdate ? Optional.of(metadata) : Optional.empty());
+  }
+
+  ////////////////////
+  // Helper methods //
+  ////////////////////
+
+  private CloseableIterator<Row> getMetadataActions() {
+    final List<Row> metadataActions = new ArrayList<>();
+
+    metadataActions.add(createCommitInfoSingleAction(commitInfo.toRow()));
+
+    if (txnState.isProtocolUpdate) {
+      metadataActions.add(createProtocolSingleAction(txnState.protocol.toRow()));
+    }
+
+    if (txnState.isMetadataUpdate) {
+      metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+    }
+
+    txnState.setTxnOpt.ifPresent(
+        setTxn -> metadataActions.add(createTxnSingleAction(setTxn.toRow())));
+
+    return toCloseableIterator(metadataActions.iterator());
+  }
+
+  private long getCommitAsVersion() {
+    return txnState.readTableOpt.map(readTable -> readTable.getVersion() + 1).orElse(0L);
+  }
+
+  private CommitInfo getCommitInfo() {
+    return new CommitInfo(
+        Optional.empty(), /* inCommitTimestampOpt */ // TODO: support ICT
+        commitAttemptTimestampMs,
+        "Kernel-" + Meta.KERNEL_VERSION + "/" + txnState.engineInfo, /* engineInfo */
+        txnState.operation.getDescription(), /* description */
+        getOperationParameters(), /* operationParameters */
+        true, /* isBlindAppend */
```

Comment:
I think we want to not hard-code this but we will revisit this when we come back to our APIs anyways

Reply:
Yeah I'm just hard-coding this for now (trying to move fast). As we allow non-blind-appends and have new and improvemed intent-based APIs, we will for sure update this 👍